cmd/bosun/sched/sched.go

```diff
@@ -704,6 +725,111 @@ func (a ActionType) MarshalJSON() ([]byte, error) {
 	return json.Marshal(a.String())
 }
 
+type Incident struct {
+	Id       uint32
+	Start    time.Time
+	End      time.Time
+	AlertKey expr.AlertKey
+}
+
+func (s *Schedule) createIncident(ak expr.AlertKey, start time.Time) *Incident {
+	s.incidentLock.Lock()
+	defer s.incidentLock.Unlock()
+	if s.Incidents == nil {
+		s.Incidents = map[uint32]*Incident{}
+	}
+	s.nextIncidentId++
+	id := s.nextIncidentId
+	incident := &Incident{
+		Id:       id,
+		Start:    start,
+		AlertKey: ak,
+	}
+
+	s.Incidents[id] = incident
+	return incident
+}
+
+func (s *Schedule) createHistoricIncidents() {
+	for ak, state := range s.status {
+		var currentIncident *Incident
+		for i, ev := range state.History {
+			if currentIncident != nil {
+				if currentIncident.End.IsZero() || ev.Time.Before(currentIncident.End) {
+					// continue open incident
+					ev.IncidentId = currentIncident.Id
+					state.History[i] = ev
+					continue
+				} else {
+					currentIncident = nil
+				}
+			}
+			if ev.Status == StNormal {
+				continue
+			}
+			// new incident
+			currentIncident = s.createIncident(ak, ev.Time)
+			ev.IncidentId = currentIncident.Id
+			state.History[i] = ev
+			// find end time for incident
+			for _, action := range state.Actions {
+				if action.Type == ActionClose && action.Time.After(ev.Time) {
+					currentIncident.End = action.Time
+					break
+				}
+			}
+
+		}
+	}
+}
+
+func (s *Schedule) GetIncidents(alert string, from, to time.Time) []*Incident {
+	s.incidentLock.Lock()
+	defer s.incidentLock.Unlock()
+	list := []*Incident{}
+	for _, i := range s.Incidents {
+		if alert != "" && i.AlertKey.Name() != alert {
+			continue
+		}
+		if i.Start.Before(from) || i.Start.After(to) {
+			continue
+		}
+		list = append(list, i)
+		if len(list) >= 200 {
+			break
+		}
+	}
+	return list
+}
+
+func (s *Schedule) GetIncidentEvents(id uint32) (*Incident, []Event, error) {
+	s.incidentLock.Lock()
+	incident, ok := s.Incidents[id]
+	s.incidentLock.Unlock()
+	if !ok {
+		return nil, nil, fmt.Errorf("Incident %d not found.", id)
+	}
+	list := []Event{}
+	state := s.GetStatus(incident.AlertKey)
+	if state == nil {
+		return incident, list, nil
+	}
+	found := false
+	for i, e := range state.History {
+		if e.IncidentId == id {
+			if !found && i > 0 {
+				list = append(list, state.History[i-1]) //add event right before incident
```

Comment:
The untention was to grab the events before and after the incident as well. After talking a bit, I don't think that is necessary anymore. Unless you think it is.